Lib/typing.py

```diff
@@ -280,25 +284,51 @@ def _collect_parameters(args):
                         parameters.append(collected)
         elif hasattr(t, '__typing_subst__'):
             if t not in parameters:
+                if type_var_tuple_encountered and t.__default__ is not None:
+                    raise TypeError('Type parameter with a default'
+                                    ' follows TypeVarTuple')
+
+                if t.__default__ is not None:
+                    default_encountered = True
+                elif default_encountered:
+                    raise TypeError(f'Type parameter {t!r} without a default'
+                                    ' follows type parameter with a default')
+
                 parameters.append(t)
         else:
+            if _is_unpacked_typevartuple(t):
+                type_var_tuple_encountered = True
             for x in getattr(t, '__parameters__', ()):
                 if x not in parameters:
                     parameters.append(x)
     return tuple(parameters)
 
 
-def _check_generic(cls, parameters, elen):
+def _check_generic_specialization(cls, parameters):
     """Check correct count for parameters of a generic cls (internal helper).
 
     This gives a nice error message in case of count mismatch.
     """
+    elen = len(cls.__parameters__)
     if not elen:
         raise TypeError(f"{cls} is not a generic class")
     alen = len(parameters)
     if alen != elen:
-        raise TypeError(f"Too {'many' if alen > elen else 'few'} arguments for {cls};"
-                        f" actual {alen}, expected {elen}")
+        # deal with defaults
+        if alen < elen:
+            # since we validate defaults in _collect_type_vars
+            # or _collect_parameters, we can safely check parameters[alen]
```

Comment:
This comment's a bit opaque right now:
- There is no `_collect_type_vars()` function (did you mean `_collect_parameters`?
- What do you mean by "we validate defaults"?
- What do you mean by "safely check"?

Maybe something like this?

```suggestion
            # A this point, we've already validated in `_collect_parameters()`
            # that any parameters with defaults appear at the end of the
            # parameters list. As such, we can guarantee here that...
```

[I'm not actually sure what we can guarantee as a result of this, you tell me 😄]

Reply:
> What do you mean by "we validate defaults"?

I think it's about validating the order (required, then optional with defaults).

> What do you mean by "safely check"?

The order being checked, you know that you have `Params = [T1, ..., Tk, D1, ..., Dk]` with `Ti` required and `Di` optional. Thus, `Params[alen].__default__` being not None means that there is a default value. So, everything that comes after has also a default. So you don't have "unfilled" values.

Reply:
I think the comment actually doesn't add much. At this point, we know that `actual_len < len(cls.__parameters__)`, so obviously `cls.__parameters__[actual_len]` is safe; that doesn't need references to other functions.

Reply:
Oh actually, the non-obvious part is that we *only* have to check the one at actual_len. If that one has a default, it implies that all later params also have a default.

Reply:
Yeah, my main point is that I find the current comment confusing; I don't feel too strongly about whether it's clarified or removed

Reply:
Pushed a reworded version